Refactor analysis base and use paralell streams

Entenwilli opened the pull request:
This PR refactors the base of the analysis and aims to improve performance by using parallel streams.
It refactors the following elements:
- Analysis initialization with a builder
- Node and variable calculation
- Sequence entities to use shared code
- Update dependency file and readme to reflect changes

This PR closes #49, #50, #51 and #53

sebinside commented:
How is the state of this PR? Ready to be reviewed?

sebinside commented on bundles/org.palladiosimulator.dataflow.confidentiality.analysis/src/org/palladiosimulator/dataflow/confidentiality/analysis/characteristics/node/PCMNodeCharacteristicsCalculatorImpl.java:
```diff
@@ -0,0 +1,277 @@
+import org.palladiosimulator.pcm.usagemodel.UsageScenario;
+import org.palladiosimulator.pcm.usagemodel.UsagemodelPackage;
+
+public class PCMNodeCharacteristicsCalculatorImpl implements NodeCharacteristicsCalculator {
```
Naming: Remove `Impl`

sebinside commented on bundles/org.palladiosimulator.dataflow.confidentiality.analysis/src/org/palladiosimulator/dataflow/confidentiality/analysis/characteristics/node/PCMNodeCharacteristicsCalculatorImpl.java:
```diff
@@ -0,0 +1,277 @@
+	            .findFirst().orElse(NodeCharacteristicsFactory.eINSTANCE.createAssignments());
+	}
+	
+	public void checkAssignments(Assignments assignments) {
```
We should discuss the public API of this class. Its super class only required one method, however, several more are public. Are they required outside this class?

sebinside commented on bundles/org.palladiosimulator.dataflow.confidentiality.analysis/src/org/palladiosimulator/dataflow/confidentiality/analysis/characteristics/variable/PCMDataCharacteristicsCalculator.java:
```diff
@@ -28,19 +26,22 @@
      */
-    public PCMDataCharacteristicsCalculator(List<DataFlowVariable> initialVariables, List<CharacteristicValue> nodeCharacteristics) {
+    public PCMDataCharacteristicsCalculator(List<DataFlowVariable> initialVariables, 
+    		List<CharacteristicValue> nodeCharacteristics,
```
I don't understand the use of node characteristics in this class, we should discuss this (and the javadoc should reflect the usage of the variable)

sebinside replied:
Document that this is because of `.container`

sebinside commented on bundles/org.palladiosimulator.dataflow.confidentiality.analysis/src/org/palladiosimulator/dataflow/confidentiality/analysis/characteristics/variable/DataCharacteristicsCalculatorFactory.java:
```diff
@@ -0,0 +1,17 @@
+import org.palladiosimulator.dataflow.confidentiality.analysis.characteristics.CharacteristicValue;
+import org.palladiosimulator.dataflow.confidentiality.analysis.characteristics.DataFlowVariable;
+
+public interface DataCharacteristicsCalculatorFactory {
```
Why is a factory required here while there is no factory required to create the node characteristics calculators?

sebinside commented on bundles/org.palladiosimulator.dataflow.confidentiality.analysis/src/org/palladiosimulator/dataflow/confidentiality/analysis/dsl/AbstractStandalonePCMDataFlowConfidentialityAnalysis.java:
```diff
@@ -0,0 +1,229 @@
+package org.palladiosimulator.dataflow.confidentiality.analysis.dsl;
```
I'm not sure about the `dsl` in the package name. I would not see this as a dsl but rather as the analysis core classes

sebinside replied:
Besides this, I like the separation in abstract, legacy and normal

sebinside commented on bundles/org.palladiosimulator.dataflow.confidentiality.analysis/src/org/palladiosimulator/dataflow/confidentiality/analysis/dsl/LegacyStandalonePCMDataFlowConfidentialityAnalysis.java:
```diff
@@ -0,0 +1,50 @@
+	 * @param relativeUsageModelPath Relative path to the usage model
+	 * @param relativeAllocationModelPath Relative path to the allocation model
+	 */
+	public LegacyStandalonePCMDataFlowConfidentialityAnalysis(PCMAnalysisBuilderData builderData, AnalysisData analysisData) {
```
The javadoc does not match the constructor

sebinside reviewed (changes requested):
First of all: This is a massive pull request with many changes to the internal analysis structure and how it is created and used. Thank you for this work, it really looks like a lot of effort.

I did not test the analysis yet, this review only contains static comments to the code. Some comments can easily be resolved in the final PR version, others should be discussed first. All in all, the refactored version should be finally what we imagined of a usable data flow analysis and I would argue that the current version of your PR is already 90% there.

sebinside commented on bundles/org.palladiosimulator.dataflow.confidentiality.analysis/src/org/palladiosimulator/dataflow/confidentiality/analysis/builder/pcm/PCMDataFlowConfidentialityAnalysisBuilder.java:
```diff
@@ -0,0 +1,104 @@
+	public AbstractStandalonePCMDataFlowConfidentialityAnalysis build() {
+		this.checkBuilderData();
+		if (this.builderData.isLegacy()) {
+			return new LegacyStandalonePCMDataFlowConfidentialityAnalysis(builderData, builderData.createAnalysisData());
```
-> As discussed, an analysis should know nothing about its builder's data

Entenwilli commented on README.md:
```diff
@@ -1,30 +1,45 @@
 # Palladio-Addons-DataFlowConfidentiality-Analysis
 
-This repository contains an analysis to extract dataflows from a palladio modelling project, analyze their data characteristics and formulate constraints to check if definied bounds are respected by the model.
+This repository contains an analysis to extract dataflows from a palladio modelling project, 
```
I've configured my editor to format like the plugin you linked, so for future changes it should be fine

Entenwilli commented:
I've addressed the requested changes and added support for loading resources directly. Now the branch should be ready. @sebinside 

Plan for future versions:
- Validation of analysis data: Assignments and resource loading
